google-cloud-spanner/src/main/java/com/google/cloud/spanner/BatchClientImpl.java

```diff
@@ -214,9 +214,14 @@ public ResultSet execute(Partition partition) throws SpannerException {
           partition.getPartitionToken());
     }
 
+    /**
+     * Closes the session as part of the cleanup. It is the responsibility of the caller to make
+     * call to this method once the transaction completes execution across all the channels (which
+     * is understandably hard to identify). It is okay if the caller does not call the method
+     * because the backend will anyways clean up the unused session.
+     */
     @Override
-    public void close() {
-      super.close();
```

Comment:
Wait, won't this break all the existing clients that are relying on the close method to return the session to the pool? (won't this possibly cause a session pool exhaustion for existing clients?)

cc @olavloite

Reply:
As BatchReadOnlyTransactionImpl is an AutoCloseable (via MultiUseReadOnlyTransaction->AbstractReadContext->ReadContext->AutoClosable), having it within a try-with-resources construct allows automatic closing of the transaction resource, which leads the underlying session to be released to the session pool. Ideally, it should not lead to session pool exhaustion.

Having an overridden close method with just super.close() won't make sense, as the super's close method will anyways be called during automatic closing.

@olavloite : Please do share your viewpoint as well.

Reply:
Oh got it, thanks for the explanation!